Short functor type syntax

lpw25 opened the pull request:
Add support for simple functor types of the form:

```
S -> T
```

equivalent to:

```
functor (_ : S) -> T
```

The `S` and `T` can be any module type expressions. 

While not actually ambiguous there is a potential shift-reduce conflict related to with constraints:

```
S with type t = int -> float
```

vs

```
S with type t = int -> T
```

This is resolved in favour of the first one (shift), meaning that the second type must be written:

```
(S with type t = int) -> T
```

yallop commented:
There are a few places in ctypes where this would make the code more succinct.  Here's one, from [cstubs.mli](https://github.com/ocamllabs/ocaml-ctypes/blob/98f4ff90ff6fbff2d23ee4a4ab0f7e37c182f2a0/src/cstubs/cstubs.mli#L66):

```ocaml
module type BINDINGS = functor (F : TYPE) -> sig end
```

Since `F` is not used in the output, we could write

```ocaml
module type BINDINGS = TYPE -> sig end
```

gasche commented:
There was a code review of this patch during the last developer meeting (I wasn't there), and the complain was the `show` boolean that incurs a lot of code changes. Remove this boolean (and simply do not print the name when it is `_`), and the patch will be ready for merging.

lpw25 committed:
Add simpler functor type syntax

Add support for simple functor types of the form:

    S -> T

equivalent to:

    functor (_ : S) -> T

lpw25 commented:
I've removed the show boolean and rebased the patch on current trunk.

gasche commented:
It's nice to see #42 finally merged in trunk! Thanks for the nice work -- and testsuite coverage.

bluddy commented:
very cool
